src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs

```diff
@@ -58,22 +62,77 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
 
         public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder) where T : IWebJobsStartup, new()
         {
-            return builder.UseWebJobsStartup(typeof(T));
+            return builder.UseWebJobsStartup<T>(NullLoggerFactory.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder, ILoggerFactory loggerFactory) where T : IWebJobsStartup, new()
+        {
+            return builder.UseWebJobsStartup(typeof(T), loggerFactory);
         }
 
         public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType)
+        {
+            return builder.UseWebJobsStartup(startupType, NullLoggerFactory.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType, ILoggerFactory loggerFactory)
         {
             if (!typeof(IWebJobsStartup).IsAssignableFrom(startupType))
             {
                 throw new ArgumentException($"The {nameof(startupType)} argument must be an implementation of {typeof(IWebJobsStartup).FullName}");
             }
 
             IWebJobsStartup startup = (IWebJobsStartup)Activator.CreateInstance(startupType);
-            startup.Configure(builder);
 
+            if (loggerFactory == NullLoggerFactory.Instance)
+            {
+                startup.Configure(builder);
+            }
+            else
+            {
+                ConfigureAndLogUserConfiguredServices(startup, builder, loggerFactory);
+            }
             return builder;
         }
 
+        private static void ConfigureAndLogUserConfiguredServices(IWebJobsStartup startup, IWebJobsBuilder builder, ILoggerFactory loggerFactory)
+        {
+            var logger = loggerFactory.CreateLogger<ServiceCollectionDecorator>();
+
+            if (builder.Services is ITrackedServiceCollection tracker)
+            {
+                if (tracker != null)
+                {
+                    startup.Configure(builder);
+                    StringBuilder sb = new StringBuilder("Services registered by external startup type " + startup.GetType().ToString() + ":");
+
+                    foreach (ServiceDescriptor service in tracker.TrackedCollectionChanges)
+                    {
+                        string serviceName = "Service name: " + service.ServiceType.FullName;
+                        StringBuilder registeredService = new StringBuilder(", Implementation: ");
```

Comment:
Some logging string nits :-)

1. Don't create a new `StringBuilder` here -- use the existing one `sb` from above.
2. Rather than storing a separate string here for `serviceName` -- just append it. It'll make it a little easier to follow. (see below)
3. Saying "Service name" on every line isn't required as we'll understand it -- and it's mostly for debugging so we can be a little short with it. (see below)
4. Rather than doing "[instance]" -- just replace "Implementation" with "Instance" on those lines. (see below)
5. Use `sb.AppendLine()` -- which I think will use the right line feed mechanics, depending on Linux/Windows. Similarly, I'd use explicit spaces rather than `\t`.

So you can start each line with with:
`sb.AppendLine($"  {service.ServiceType.FullName}: ");` (note just 2 spaces here)

Then depending on whether Implemention/Instance/Factory:
`sb.Append($"Implementation: {service.ImplementationType.FullName}");`
...
`sb.Append($"Instance: {service.ImplementationInstance.GetType().FullName}");`
...
`sb.Append("Factory");`

Reply:
I think AppendLine adds a newline to the end of the string, so could use with the first and last strings. Otherwise, I have addressed the rest of the comments. Thanks Brett!